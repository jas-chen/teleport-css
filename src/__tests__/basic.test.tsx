import { expect, test } from 'vitest';
import { renderToStaticMarkup } from 'react-dom/server';
import { MouseEvent } from 'react';
import fnv1a from '@sindresorhus/fnv1a';
import { create } from '..';
import pretty from 'pretty';

function hashFn(value: string) {
  return fnv1a(value, { size: 64 }).toString(36).slice(0, 8);
}

const { styled } = create({
  hashFn,
});

test('simple', () => {
  const Button = styled('button', () => ({
    color: 'red',
    background: 'blue',
  }));

  const Link = styled('a', () => ({
    color: 'yellow',
    background: 'blue',
  }));

  expect(
    pretty(
      renderToStaticMarkup(
        <div>
          <Button
            onClick={
              // test typing
              (e: MouseEvent<HTMLButtonElement>) => {
                console.log(e);
              }
            }
          />
          <Link />
        </div>,
      ),
    ),
  ).toMatchSnapshot();
});

test('no footgun', () => {
  const Last = styled('span', () => ({
    '&:first-child:after': {
      content: '"First"',
    },
    '&:last-child:after': {
      content: '"Last"',
    },
  }));

  const First = styled('span', () => ({
    '&:last-child:after': {
      content: '"Last"',
    },
    '&:first-child:after': {
      content: '"First"',
    },
  }));

  expect(
    pretty(
      renderToStaticMarkup(
        <div>
          <Last />
          <First />
        </div>,
      ),
    ),
  ).toMatchSnapshot();
});

test('CSS value fallback', () => {
  const Text = styled('span', () => ({
    color: 'rgba(0,0,0,0.9)',
  }));

  const TextWithFallback = styled('span', () => [
    {
      color: 'black;color:rgba(0,0,0,0.9)',
    },
  ]);

  expect(
    pretty(
      renderToStaticMarkup(
        <div>
          <TextWithFallback />
          <Text />
        </div>,
      ),
    ),
  ).toMatchSnapshot();
});

test('style composition using styled', () => {
  const Button = styled('button', () => ({
    color: 'red',
    background: '#fff',
  }));

  const PrimaryButton = styled(Button, () => ({
    background: 'blue',
  }));

  expect(
    pretty(
      renderToStaticMarkup(
        <div>
          <Button
            onClick={
              // test typing
              (e: MouseEvent<HTMLButtonElement>) => {
                console.log(e);
              }
            }
          />
          <PrimaryButton
            onClick={
              // test typing
              (e: MouseEvent<HTMLButtonElement>) => {
                console.log(e);
              }
            }
          />
        </div>,
      ),
    ),
  ).toMatchSnapshot();
});

test('style composition using css - functions', () => {
  const Button = styled('button', () => ({
    background: 'white',
  }));

  expect(
    pretty(
      renderToStaticMarkup(
        <div>
          <Button
            css={() => ({
              background: 'red',
            })}
          />
          <Button
            css={() => ({
              background: 'blue',
            })}
          />
          <Button />
        </div>,
      ),
    ),
  ).toMatchSnapshot();
});

test('style composition using css - objects', () => {
  const Button = styled('button', () => ({
    background: 'white',
  }));

  expect(
    pretty(
      renderToStaticMarkup(
        <div>
          <Button
            css={{
              background: 'red',
            }}
          />
          <Button
            css={[
              {
                background: 'blue',
              },
            ]}
          />
          <Button />
        </div>,
      ),
    ),
  ).toMatchSnapshot();
});
