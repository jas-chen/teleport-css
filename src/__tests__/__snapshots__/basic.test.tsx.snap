// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CSS value fallback 1`] = `
"<style data-precedence="1" data-href="x-3snbh6it-1 x-ggil3886-1">
  .x-3snbh6it-1 {
    color: black
  }

  .x-ggil3886-1 {
    color: color: rgba(0, 0, 0, 0.9)
  }
</style>
<style data-precedence="2" data-href="x-ggil3886-2">
  .x-ggil3886-2 {
    color: color: rgba(0, 0, 0, 0.9)
  }
</style>
<div><span class="x-3snbh6it-1 x-ggil3886-2"></span><span class="x-ggil3886-1"></span></div>"
`;

exports[`no footgun 1`] = `
"<style data-precedence="1" data-href="x-9p6p2vms-1 x-16x8ijtr-1">
  .x-9p6p2vms-1 {
    &:first-child:after {
      content: "First"
    }
  }

  .x-16x8ijtr-1 {
    &:last-child:after {
      content: "Last"
    }
  }
</style>
<style data-precedence="2" data-href="x-16x8ijtr-2 x-9p6p2vms-2">
  .x-16x8ijtr-2 {
    &:last-child:after {
      content: "Last"
    }
  }

  .x-9p6p2vms-2 {
    &:first-child:after {
      content: "First"
    }
  }
</style>
<div><span class="x-9p6p2vms-1 x-16x8ijtr-2"></span><span class="x-16x8ijtr-1 x-9p6p2vms-2"></span></div>"
`;

exports[`style composition 1`] = `
"<style data-precedence="1" data-href="x-157mpepm-1 x-mqt2cko5-1">
  .x-157mpepm-1 {
    color: red
  }

  .x-mqt2cko5-1 {
    background: #fff
  }
</style>
<style data-precedence="2" data-href="x-39u6vegt-2">
  .x-39u6vegt-2 {
    background: blue
  }
</style>
<div><button class="x-157mpepm-1 x-mqt2cko5-1"></button><button class="x-157mpepm-1 x-mqt2cko5-1 x-39u6vegt-2"></button></div>"
`;

exports[`style deduplication 1`] = `
"<style data-precedence="1" data-href="x-157mpepm-1 x-39u6vegt-1 x-38xvzlwd-1">
  .x-157mpepm-1 {
    color: red
  }

  .x-39u6vegt-1 {
    background: blue
  }

  .x-38xvzlwd-1 {
    color: yellow
  }
</style>
<div><button class="x-157mpepm-1 x-39u6vegt-1"></button><a class="x-38xvzlwd-1 x-39u6vegt-1"></a></div>"
`;
